lib/workflow/stores/mongo.js

```diff
@@ -643,18 +643,34 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             graphId: data.graphId,
             reachable: true
         };
-        query['dependencies.' + data.taskId] = {
-            $in: [data.state, Constants.Task.States.Finished]
-        };
         var update = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
         var options = {
             multi: true
         };
+        
+        query['dependencies.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
 
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        return waterline.taskdependencies.find(query).then(function(res) {
+           if (_.isEmpty(res)) {
+                // no task found under dependencies try search key anyOf
+                delete query['dependencies.' + data.taskId];
+                query['dependencies.anyOf.' + data.taskId] = {
+                    $in: [data.state, Constants.Task.States.Finished]
+                };
+
+                update.$unset['dependencies.anyOf'] = '';
+                
+                return waterline.taskdependencies.updateMongo(query, update, options);
+            }
+            else {
```

Comment:
To minimize the back-and-forth database operation and thus improve the performance after enabling the handling for above two cases, you could read the whole `dependencies` into memory (using OR query), manipulate the dependencies in the memory,  then write it back to database.